### PR TITLE
add test for use pvc to expand volume (#2422)

### PR DESCRIPTION
**Pull request by cclhsu**
add test for use pvc to expand volume to resolve ticket 'Application couldn't mount the volume after expanded the volume' (#2422)

Signed-off-by: Clark Hsu <clark.hsu@suse.com>

### CODE REVIEW

**Comment by khushboo-rancher** [comment]
We can add this in v1.1.2 as it is back ported to v1.1.2.

File: docs/content/manual/release-specific/v1.2.0/expand.md [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        Given Kubernetes cluster deployed
+        And Longhorn deployed
+        And a statefulset/deployment deployed with mounted volume
+```
+
+```Gherkin
+    Scenario: statefulset/deployment is replica is zero and expand with Longhorn API
+        Given Background conditions met
+        When scale down replicas for statefulset/deployment to zero
+            """bash
```
Comment (khushboo-rancher) [inline docs/content/manual/release-specific/v1.2.0/expand.md:29]: In rich view, `"""bash` is appearing as it is.

File: manager/integration/tests/test_statefulset.py [hunk 2]
```diff
@@ -438,3 +439,133 @@ def test_statefulset_restore(set_random_backupstore, client, core_api, storage_c
...
+
+
+def test_statefulset_volume_expand(client, core_api, storage_class, statefulset): # NOQA
+    """
+    Test that expanded volume can be mounted by a StatefulSet.
+
+    1. Create a StatefulSet with VolumeClaimTemplate and Longhorn.
+    2. Wait for pods to run.
+    3. Scale down replicas to 0
+    4. Expand volume using pvc
```
Comment (shuo-wu) [inline manager/integration/tests/test_statefulset.py:451]: This case is for testing the update HTTP API `Expand`. You need to do expansion via `v.expand()` then check if both PVC and the volume itself get updated.

**Review by shuo-wu — CHANGES_REQUESTED** [review]

**Comment by innobead** [comment]
Can you use the complete name for the test suite file? `expaned.md` is not a good name.. you can refer to existing examples.

**Comment by innobead** [comment]
If your test cases can be covered in automated test cases, I don't think you need to create duplicated test cases for manual regression. 

Your test cases in expand.md seems not really focus on what you changed like why you emphaisze replica is zero or not zero? I would suggest you just need to cover below senarios.

- Successful/Failed volume expansion w/o PVC
- Successful/Failed volume expansion w/ PVC
- Also check the round value of volume size (this is what you changed as well) based on the above cases

Workload types should do not really matter if u know the behavior, because this is more like white box testing already.

**Review by innobead — CHANGES_REQUESTED** [review]
Please update based on my suggestions.

**Comment by cclhsu** [comment]
> If your test cases can be covered in automated test cases, I don't think you need to create duplicated test cases for manual regression.
> 
> Your test cases in expand.md seems not really focus on what you changed like why you emphaisze replica is zero or not zero? I would suggest you just need to cover below senarios.
> 
> * Successful/Failed volume expansion w/o PVC
> * Successful/Failed volume expansion w/ PVC
> * Also check the round value of volume size (this is what you changed as well) based on the above cases
> 
> Workload types should do not really matter if u know the behavior, because this is more like white box testing already.

Change the test cases accordingly.  The new test cases result is in https://ci.longhorn.io/job/private/job/longhorn-tests-regression/241/testReport/tests/test_basic/.

File: manager/integration/tests/test_basic.py [hunk 3]
```diff
@@ -3658,3 +3660,128 @@ def test_volume_toomanysnapshots_condition(client, core_api, volume_name): # NOQ
...
+    volume.attach(hostId=self_hostId, disableFrontend=False)
+    wait_for_volume_healthy(client, volume_name)
+    volume.detach(hostId="")
+    volume = wait_for_volume_detached(client, volume_name)
+
+    client.delete(volume)
+    wait_for_volume_delete(client, volume_name)
+
+
+def test_expand_pvc(client, core_api, volume_name, pod):  # NOQA
```
Comment (shuo-wu) [inline manager/integration/tests/test_basic.py:3711]: `test_csi_expansion_with_size_round_up`
And you can put this test in `test_csi.py`

File: manager/integration/tests/test_basic.py [hunk 4]
```diff
@@ -3658,3 +3660,128 @@ def test_volume_toomanysnapshots_condition(client, core_api, volume_name): # NOQ
...
+    """
+    test expand lhv volume
```
Comment (shuo-wu) [inline manager/integration/tests/test_basic.py:3713]: expression `lhv volume` is weird. `lhv` is the alias of `longhorn volume`.

File: manager/integration/tests/test_basic.py [hunk 5]
```diff
@@ -3658,3 +3660,128 @@ def test_volume_toomanysnapshots_condition(client, core_api, volume_name): # NOQ
...
+
+    1. Create LHV,PV,PVC volume with size '1Gi'
+    2. attach and detach
```
Comment (shuo-wu) [inline manager/integration/tests/test_basic.py:3716]: Attach and mount the volume, write some data into the mount point, then detach the volume.

**Review by shuo-wu — CHANGES_REQUESTED** [review]
1. The target of this feature is asking Longhorn to automatically expand PVC via the HTTP API `expand`. Do we need `test_expand_lhv`? cc @innobead 
2. IIRC, the expansion API (both in HTTP and CSI) won't round up the size. Do we need a manager PR to fix this? Anyway, it's great to test this here.

File: manager/integration/tests/test_basic.py (hunk 3 above)
Comment (cclhsu) [reply manager/integration/tests/test_basic.py]: changed

File: manager/integration/tests/test_basic.py (hunk 4 above)
Comment (cclhsu) [reply manager/integration/tests/test_basic.py]: changed

File: manager/integration/tests/test_basic.py (hunk 5 above)
Comment (cclhsu) [reply manager/integration/tests/test_basic.py]: added

File: manager/integration/tests/test_basic.py [hunk 6]
```diff
@@ -3658,3 +3660,86 @@ def test_volume_toomanysnapshots_condition(client, core_api, volume_name): # NOQ
...
+
+    volume = client.by_id_volume(volume_name)
+    assert volume.size == "2147483648"
+    volume.detach(hostId="")
+    volume = wait_for_volume_detached(client, volume_name)
+
+    self_hostId = get_self_host_id()
+    volume.attach(hostId=self_hostId, disableFrontend=False)
+    volume = wait_for_volume_healthy(client, volume_name)
+    write_volume_random_data(volume)
```
Comment (shuo-wu) [inline manager/integration/tests/test_basic.py:3740]: Need to verify existing data first.

Same below

**Review by shuo-wu — CHANGES_REQUESTED** [review]

File: manager/integration/tests/test_basic.py (hunk 6 above)
Comment (cclhsu) [reply manager/integration/tests/test_basic.py]: changed

**Review by shuo-wu — APPROVED** [review]

**Pull request merged**